File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1] * (len(nums) + 1)
        for i in range(len(nums)):
            prefix[i+1] = prefix[i] * nums[i]
        prefix = prefix[1:]
        postfix = [1] * (len(nums) + 1)
        i = len(nums) - 1
        while i > -1:
            postfix[i] = postfix[i+1] * nums[i]
            i -= 1
        postfix = postfix[:len(postfix)-1]
        res = [1] * len(nums)
        for i in range(len(nums)):
            if i == 0:
                res[i] = postfix[i+1]
            elif i == len(nums) - 1:
                res[i] = prefix[i-1]
            else:
                res[i] = postfix[i+1] * prefix[i-1]
        return res
        

File: Data Structures & Algorithms/kth-largest-element-in-an-array/submission-0.go
type MaxHeap []int

func (h *MaxHeap) Push(a interface{}) {
	*h = append(*h, a.(int))
}

func (h *MaxHeap) Pop() interface{} {
	x := (*h)[len(*h) - 1]
	*h = (*h)[:len(*h) - 1]
	return x
}

func (h MaxHeap) Len() int {
	return len(h)
}

func (h MaxHeap) Less(i, j int) bool {
	return h[i] > h[j]
}

func (h MaxHeap) Swap(i,j int) {
	h[i], h[j] = h[j], h[i]
}

func findKthLargest(nums []int, k int) int {
	h := &MaxHeap{}
	heap.Init(h)
	for _, el := range nums {
		heap.Push(h, el)
	}
	res := heap.Pop(h).(int)
	for i := 1 ; i < k ; i++ {
		res = heap.Pop(h).(int)
	}
	return res
}
